test/Test_Menhir.re

```diff
@@ -0,0 +1,860 @@
+ * serialize it, parse it with MakeTerm, and parse it with Menhir.
+ */
+let qcheck_menhir_maketerm_equivalent_test =
+  QCheck.Test.make(
+    ~name="Menhir and maketerm are equivalent",
+    ~count=100,
+    QCheck.make(~print=AST.show_exp, AST.gen_exp_sized(7)),
+    exp => {
+      let core_exp = Conversion.Exp.of_menhir_ast(exp);
+      // TODO Maybe only do this when necessary.
+      // TODO Check with Matthew if I'm using this correctly
+      // Add parens around tuples and parens around certain patterns
+      // Move this to conversion
+      let core_exp =
+        Exp.map_term(
+          ~f_exp=
+            (cont, e) =>
+              switch (e.term) {
+              | Tuple(es) =>
+                Parens(Tuple(es |> List.map(cont)) |> Exp.fresh)
+                |> Exp.fresh
+              | _ => cont(e)
+              },
+          ~f_pat=
+            (cont, e) =>
+              switch (e.term) {
+              | Tuple(es) =>
+                Parens(Tuple(es |> List.map(cont)) |> Pat.fresh)
+                |> Pat.fresh
+              | Cons(p1, p2) =>
+                Parens(Cons(cont(p1), cont(p2)) |> Pat.fresh) |> Pat.fresh
+              | Cast(p, t1, t2) =>
+                Parens(Cast(cont(p), t1, t2) |> Pat.fresh) |> Pat.fresh
+              | _ => cont(e)
+              },
+          ~f_typ=
+            (cont, e) =>
+              switch (e.term) {
+              | Rec(x, t) =>
+                Parens(Rec(x, cont(t)) |> Typ.fresh) |> Typ.fresh
+              | Sum(cs) => Parens(Sum(cs) |> Typ.fresh) |> Typ.fresh
+              | Prod(ts) => Parens(Prod(ts) |> Typ.fresh) |> Typ.fresh
+              | _ => cont(e)
+              },
+          core_exp,
+        );
+
+      let segment =
+        ExpToSegment.exp_to_segment(
+          ~settings=
+            ExpToSegment.Settings.of_core(
+              ~inline=true, // TODO What does inline do?
+              CoreSettings.off,
+            ),
+          core_exp,
+        );
+
+      let serialized = Printer.of_segment(~holes=Some("?"), segment);
+      let make_term_parsed = make_term_parse(serialized);
+      let menhir_parsed = Haz3lmenhir.Interface.parse_program(serialized);
+      let menhir_parsed_converted =
+        Haz3lmenhir.Conversion.Exp.of_menhir_ast(menhir_parsed);
+
+      switch (
+        Haz3lcore.DHExp.fast_equal(make_term_parsed, menhir_parsed_converted)
+      ) {
+      | true => true
+      | false => false
+      | exception (Failure(msg)) =>
+        print_endline("Error: " ++ msg);
+        msg == "Sum type has non-unique constructors";
+      };
+    },
+  );
+
+/**
+ * QCheck Test to check that menhir parses out what ExpToSegment serializes.
+ * We generate an expression, convert it to the core representation, convert it to a segment,
+ * serialize it, parse it with Menhir, and compare to the original.
+ *
+ * TODO This fails due to types not being serialized on constructors
+ *  and some other ExpToSegment inconsistencies
```

Comment:
file bugs

Reply:
@cyrus- these are the bugs I've found with the qcheck test. 
-  https://github.com/hazelgrove/hazel/issues/1452
-  https://github.com/hazelgrove/hazel/issues/1451
-  https://github.com/hazelgrove/hazel/issues/1445